object_store/src/local.rs

```diff
@@ -678,6 +618,94 @@ impl ObjectStore for LocalFileSystem {
     }
 }
 
+impl LocalFileSystem {
+
+    fn list_with_maybe_offset(&self, prefix: Option<&Path>, maybe_offset: Option<&Path>) -> BoxStream<'static, Result<ObjectMeta>> {
+
+        let config = Arc::clone(&self.config);
+
+        let root_path = match prefix {
+            Some(prefix) => match config.prefix_to_filesystem(prefix) {
+                Ok(path) => path,
+                Err(e) => return futures::future::ready(Err(e)).into_stream().boxed(),
+            },
+            None => config.root.to_file_path().unwrap(),
+        };
+
+        let walkdir = WalkDir::new(root_path)
+            // Don't include the root directory itself
+            .min_depth(1)
+            .follow_links(true);
+
+        let maybe_offset_clone = maybe_offset.map(move |o| o.clone());
+
+        let s = walkdir.into_iter().flat_map(move |result_dir_entry| {
+            // Apply offset filter before proceeding, to reduce statx file system calls
+            // This matters for NFS mounts
+            if maybe_offset_clone.as_ref().is_some() && result_dir_entry.is_ok() {
+                let offset = maybe_offset_clone.as_ref().unwrap();
+                let entry = result_dir_entry.as_ref().unwrap();
+                let location = config.filesystem_to_path(entry.path());
+
+                if location.unwrap() <= *offset {
```

Comment:
Could this unwrap fail? We might want to handle an Err here and return it rather than allow it to panic. It looks like `config.filesystem_to_path(entry.path())` is called further below so we might want to fix that duplication too.

Reply:
Good point. I have added error handling like what is done in the code below. I've left in the duplication since they are doing slightly different things.